### PR TITLE
Provide back button navigation in new cluster wizard

### LINKED ISSUE #21: Provide back button navigation in new cluster wizard
For instance, if I start creating a new cluster, after selecting the broker urls and the friendly name, when about to choose the authentication, I changed my mind on the friendly name, need to go back to modify it, a back button would be useful.
See [vscode-quarkus's new project wizard](https://github.com/redhat-developer/vscode-quarkus/blob/master/src/wizards/generateProject/generationWizard.ts):

Nov-23-2020 09-42-17

### PR DESCRIPTION
**Pull request by angelozerr**
Provide back button navigation in new cluster wizard

Fixes #21

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR improves the two wizards:

 * create cluster
 * add topic

with:

 * back button
 * validation
 * default value

for add topic, if you execute add topic from command palette it provides list of cluster to pick it at first.

Here a demo with cluster wizard:

AddClusterDemo

Here a demo with topic wizard:

AddTopicDemo

I need to improve this PR:

 * with topic name validation (to have the same validation than https://github.com/apache/kafka/blob/8007211cc982d8458223e866c1ee7d94b69e0249/core/src/main/scala/kafka/common/Topic.scala#L33
 * re-read my code to clean it if it is required
 * understand why the PR is failing ?

**Comment by angelozerr** [comment]
@jlandersen my PR fails on windows only when tests are executed, I don't understand why although I have not written some tests 
https://github.com/jlandersen/vscode-kafka/pull/60/checks?check_run_id=1644491781

Have you some idea?

**Comment by fbricon** [comment]
Use simple messages like "X is required"

Check for blank values: "X can not be blank"

File: src/wizards/cluster/validateInput.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+export async function validateAuthentificationUserName(userName: string): Promise<string | undefined> {
+    if (userName.length <= 0) {
+        return "user name is illegal, can't be empty";
+    }
+    return undefined;
+}
+export async function validateAuthentificationPassword(password: string): Promise<string | undefined> {
+    if (password.length <= 0) {
```
Comment (fbricon) [inline src/wizards/cluster/validateInput.ts:25]: some passwords can be empty I guess

File: src/wizards/topics/validateInput.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    else if (topic == "." || topic == "..") {
+        return "topic name cannot be \".\" or \"..\"";
+    }
+    else if (topic.length > maxNameLength) {
+        return `topic name is illegal, can't be longer than ${maxNameLength} characters`;
+    }
+    else if (!/^[a-zA-Z0-9\\.\\-]*$/.test(topic)) {
+        return "topic name " + topic + " is illegal,  contains a character other than ASCII alphanumerics, '.', '_' and '-'";
+    } else if (existingTopicNames.indexOf(topic) != -1) {
+        return `The topic '${topic}' already exists for the cluster`;
```
Comment (fbricon) [inline src/wizards/topics/validateInput.ts:17]: Topic '${topic}' already exists

**Comment by angelozerr** [comment]
> Use simple messages like "X is required"

> Check for blank values: "X can not be blank"

Done, I updated my PR and now we have a validators.ts which provides commons validator for required filed, positive number etc which follows the same pattern message.

File: src/wizards/topics/validateInput.ts (hunk 2 above)
Comment (angelozerr) [reply src/wizards/topics/validateInput.ts]: fixed

File: src/wizards/cluster/validateInput.ts (hunk 1 above)
Comment (angelozerr) [reply src/wizards/cluster/validateInput.ts]: fixed

File: src/wizards/validators.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    return validateFieldUniqueValue('Topic', topic, existingTopicNames);
+}
+
+export async function validatePartitions(partitions: string): Promise<string | undefined> {
+    const result = validateFieldRequired('Partitions', partitions);
+    if (result) {
+        return result;
+    }
+    return validateFieldPositiveNumber('Partitions', partitions);
+}
```
Comment (fbricon) [inline src/wizards/validators.ts:48]: We should prolly limit the max value for the number of topic partitions. Looks like 10 seems to be the max recommended amount, So setting [100](https://medium.com/swlh/choosing-right-partition-count-replication-factor-apache-kafka-cf50b1bc75cf) as a limit might be fine. @jlandersen WDYT?

Kafka clusters seem limited to [200K partions overall](https://blogs.apache.org/kafka/entry/apache-kafka-supports-more-partitions#:~:text=Apache%20Kafka%20Supports%20200K%20Partitions%20Per%20Cluster%20%3A%20Kafka), if anything that should be the hard limit.

File: src/wizards/multiStepInput.ts [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                        (item as QuickInputButtonWithCallback).callback();
+                        // resolve();
+                    } else if (item === QuickInputButtons.Back) {
+                        reject(InputFlowAction.back);
+                    } else {
+                        resolve(<any>item);
+                    }
+                }),
+                input.onDidAccept(async () => {
+                    const value = input.value;
```
Comment (fbricon) [inline src/wizards/multiStepInput.ts:213]: we should prolly do const value = input.value.trim();

File: src/wizards/validators.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+export async function validateReplicationFactor(replicationFactor: string): Promise<string | undefined> {
+    return validateFieldPositiveNumber('Replication Factor', replicationFactor);
```
Comment (fbricon) [inline src/wizards/validators.ts:51]: apparently you can't use a replication factor higher than the cluster's `offsets.topic.replication.factor` value.

IMHO if `offsets.topic.replication.factor`=1, then there's no point in even showing that part of the wizard.

See https://github.com/jlandersen/vscode-kafka/blob/824a615fd5d49e2021b485d694dc9706429bc1a6/src/commands/topics.ts#L145-L156 as an example to get the cluster's config value:

**Review by fbricon — CHANGES_REQUESTED** [review]

File: src/wizards/validators.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    return validateFieldUniqueValue('Topic', topic, existingTopicNames);
+}
+
+export async function validatePartitions(partitions: string): Promise<string | undefined> {
+    const result = validateFieldRequired('Partitions', partitions);
+    if (result) {
+        return result;
+    }
+    return validateFieldPositiveNumber('Partitions', partitions);
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/wizards/validators.ts]: fixed

File: src/wizards/validators.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// ------------------ Topic validators
+
+const maxNameLength = 249
+const legalChars = /^[a-zA-Z0-9\\.\\-]*$/;
+
+export async function validateTopicName(topic: string, existingTopicNames: string[]): Promise<string | undefined> {
+    // See topic name validation rule at https://github.com/apache/kafka/blob/8007211cc982d8458223e866c1ee7d94b69e0249/core/src/main/scala/kafka/common/Topic.scala#L33
+    const result = validateFieldRequired('Topic', topic);
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/wizards/validators.ts]: fixed

File: src/wizards/validators.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    else if (topic == "." || topic == "..") {
+        return "Topic cannot be \".\" or \"..\"";
+    }
+    else if (topic.length > maxNameLength) {
+        return `Topic is illegal, can't be longer than ${maxNameLength} characters`;
+    }
+    else if (!legalChars.test(topic)) {
+        return `Topic ${topic} is illegal, contains a character other than ASCII alphanumerics, '.', '_' and '-'`;
+    }
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/wizards/validators.ts]: fixed

File: src/wizards/validators.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    // See topic name validation rule at https://github.com/apache/kafka/blob/8007211cc982d8458223e866c1ee7d94b69e0249/core/src/main/scala/kafka/common/Topic.scala#L33
+    const result = validateFieldRequired(TOPIC_FIELD, topic);
+    if (result) {
+        return result;
+    }
+    else if (topic == "." || topic == "..") {
+        return `${TOPIC_FIELD} cannot be '.' or '..'`;
+    }
+    else if (topic.length > maxNameLength) {
+        return `${TOPIC_FIELD} is illegal, can't be longer than ${maxNameLength} characters`;
```
Comment (fbricon) [inline src/wizards/validators.ts:42]: need to be consistent. Use cannot throughout the validation (and rest of the codebase)

Comment (angelozerr) [reply src/wizards/validators.ts:42]: I'm sorry @fbricon I don't understand what you mean? Which fix you want I do?

Comment (fbricon) [reply src/wizards/validators.ts:42]: replace all `can't` occurrences with `cannot`

Comment (angelozerr) [reply src/wizards/validators.ts]: Ok thanks for your feedback. Fixed.

File: src/wizards/clusters.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        if (!state.saslOption) {
+            return;
+        }
+        state.saslOption.password = await input.showInputBox({
+            title: INPUT_TITLE,
+            step: input.getStepNumber(),
+            totalSteps: state.totalSteps,
+            value: state.saslOption.password || '',
+            prompt: ' Password'
+        });
```
Comment (fbricon) [inline src/wizards/clusters.ts:97]: Password appears in clear text

Comment (angelozerr) [reply src/wizards/clusters.ts:97]: Is it really a problem?

Comment (fbricon) [reply src/wizards/clusters.ts:97]: it is when performing/recording demos ;-)

Comment (angelozerr) [reply src/wizards/clusters.ts:97]: I understand that but is it possible to see the typed password?

Comment (angelozerr) [reply src/wizards/clusters.ts]: fixed

File: src/wizards/validators.ts (hunk 5 above)
Comment (angelozerr) [reply src/wizards/validators.ts]: > apparently you can't use a replication factor higher than the cluster's offsets.topic.replication.factor value.

have you a link about that?

Comment (angelozerr) [reply src/wizards/validators.ts]: See https://github.com/jlandersen/vscode-kafka/issues/64

Comment (fbricon) [reply src/wizards/validators.ts]: I dont have a link. This is what I inferred from experience.

**Comment by fbricon** [comment]
Name validation is wrong. `_` should be accepted (as explained in the error message)

File: src/wizards/validators.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+// ------------------ Topic validators
+
+const TOPIC_FIELD = 'Topic name';
+const PARTITIONS_FIELD = 'Number of partitions';
+const REPLICATION_FACTOR_FIELD = 'Replication Factor';
+
+const maxNameLength = 249
+const legalChars = /^[a-zA-Z0-9\\.\\-]*$/;
```
Comment (fbricon) [inline src/wizards/validators.ts:30]: `[a-zA-Z0-9\\._\\-]`

Comment (angelozerr) [reply src/wizards/validators.ts]: Good catch! fixed

**Comment by fbricon** [comment]
Thanks @angelozerr. This makes the wizard workflows much better. Now on to #64 and #65 to improve validation

**Pull request merged**